flash/core/data/input_transform.py

```diff
@@ -406,17 +284,755 @@ def _create_collate_input_transform_processors(self) -> Tuple[Any]:
+
+    @partial(transform_context, current_fn="collate")
+    def _collate(self, samples: Sequence, metadata=None) -> Any:
+        """Transform to convert a sequence of samples to a collated batch."""
+        current_transform = self.current_transform
+        if current_transform is self._identity:
+            current_transform = default_collate
+
+        # the model can provide a custom ``collate_fn``.
+        collate_fn = self.get_state(CollateFn)
+        if collate_fn is not None:
+            collate_fn = collate_fn.collate_fn
+        else:
+            collate_fn = current_transform
+            # return collate_fn.collate_fn(samples)
+
+        parameters = inspect.signature(collate_fn).parameters
+        if len(parameters) > 1 and DataKeys.METADATA in parameters:
+            return collate_fn(samples, metadata)
+        return collate_fn(samples)
+
+    @partial(transform_context, current_fn="per_sample_transform_on_device")
+    def _per_sample_transform_on_device(self, sample: Any) -> Any:
+        """Transforms to apply to the data before the collation (per-sample basis).
+
+        .. note::     This option is mutually exclusive with :meth:`per_batch_transform`,     since if both are
+        specified, uncollation has to be applied. .. note::     This function won't be called within the dataloader
+        workers, since to make that happen     each of the workers would have to create it's own CUDA-context which
+        would pollute GPU memory (if on GPU).
+        """
+        if isinstance(sample, list):
+            return [self.current_transform(s) for s in sample]
+        return self.current_transform(sample)
+
+    @partial(transform_context, current_fn="per_batch_transform_on_device")
+    def _per_batch_transform_on_device(self, batch: Any) -> Any:
+        """Transforms to apply to a whole batch (if possible use this for efficiency).
+
+        .. note::     This function won't be called within the dataloader workers, since to make that happen     each of
+        the workers would have to create it's own CUDA-context which would pollute GPU memory (if on GPU).
+        """
+        return self.current_transform(batch)
+
+    ########################
+    # PER SAMPLE TRANSFORM #
+    ########################
+
+    def per_sample_transform(self) -> Callable:
+        """Defines the transform to be applied on a single sample on cpu for all stages stage.
+
+        The input data of the transform would have the following form:
+        {
+            "input": ...,
+            "target": ...,
+            "metadata": ...,
+        }
+
+        You would need to use :class:`flash.core.data.transforms.ApplyToKeys` as follows:
+
+        Example:
+
+            from flash.core.data.transforms import ApplyToKeys
+
+            class MyInputTransform(InputTransform):
+
+                def per_sample_transform(self) -> Callable:
+
+                    return ApplyToKeys("input", my_func)
```

Comment:
I thought you didn't want to return this class explicitly

if you do, is the typing return wrong?

Reply:
I imagine some users might want to have the entire dictionary if there do a transformation on both input and target at the same time.